docs/design/resource-qos.md

```diff
@@ -0,0 +1,254 @@
+
+The relationship between "Requests and Limits" and "QoS Classes" is subtle. Theoretically, the policy of classifying pods into QoS classes is orthogonal to the requests and limits specified for the container. Hypothetically, users could use an (currently unplanned) API to specify whether a pod is guaranteed or best-effort. However, in the current design, the policy of classifying pods into QoS classes is intimately tied to "Requests and Limits" - in fact, QoS classes are used to implement some of the memory guarantees described in the previous section.
+
+Pods can be of one of 3 different classes:
+
+- If `limits` and optionally *requests* (not equal to `0`) are set for all resources across all containers and they are *equal*, then the container is classified as **Guaranteed**.
+
+Examples:
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				cpu: 10m
+				memory: 1Gi
+	name: bar
+		resources:
+			limits:
+				cpu: 100m
+				memory: 100Mi
+```
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				cpu: 10m
+				memory: 1Gi
+			requests:
+				cpu: 10m
+				memory: 1Gi
+
+	name: bar
+		resources:
+			limits:
+				cpu: 100m
+				memory: 100Mi
+			requests:
+				cpu: 10m
+				memory: 1Gi
+```
+
+- If `requests` and optionally `limits` are set (not equal to `0`) for one or more resources across one or more containers, and they are *not equal*, then the pod is classified as **Burstable**.
+When `limits` are not specified, they default to the node capacity.
+
+Examples:
+
+Container `bar` has not resources specified.
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				cpu: 10m
+				memory: 1Gi
+			requests:
+				cpu: 10m
+				memory: 1Gi
+
+	name: bar
+```
+
+Container `foo` and `bar` have limits set for different resources.
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				memory: 1Gi
+
+	name: bar
+		resources:
+			limits:
+				cpu: 100m
+```
+
+Container `foo` has no limits set, and `bar` has neither requests nor limits specified.
+
+```yaml
+containers:
+	name: foo
+		resources:
+			requests:
+				cpu: 10m
+				memory: 1Gi
+
+	name: bar
+```
+
+- If `requests` and `limits` are not set for all of the resources, across all containers, then the pod is classified as **Best-Effort**.
+
+Examples:
+
+```yaml
+containers:
+	name: foo
+		resources:
+	name: bar
+		resources:
+```
+
+Pods will not be killed if CPU guarantees cannot be met (for example if system tasks or daemons take up lots of CPU), they will be temporarily throttled.
+
+Memory is an incompressible resource and so let's discuss the semantics of memory management a bit.
+
+- *Best-Effort* pods will be treated as lowest priority. Processes in these pods are the first to get killed if the system runs out of memory.
```

Comment:
This is not truly accurate today.  A BestEffort pod that sets a memory limit is actually given a guaranteed oom_score_adj value, we will need to address this.  With eviction enabled, this description is accurate though, but in that case evicted means failed.